### ISSUE TITLE
Question about softmaxloss

### ISSUE THREAD

**Issue by EdwardMao** [UKPLab/sentence-transformers#17]
Hi,
This is a wonderful project. But I have a question about softmaxloss. If you train with softmax loss and when you evaluate the model, you still need the classifier layer. It means that finding in a collection of n = 10 000 sentences the pair with the highest similarity requires, you still need n·(n−1)/2 = 49 995 000 inference computations, right? Because you still need the weights in classifier layer.
I believe that in order to use cosine similarity to evaluate the model, you have to train with cosine similarityloss. Is it possible to use cosine similarity to evaluate the model after training with softmaxloss? 
Thanks

**Comment by nreimers** [comment]
Hi @EdwardMao 
thanks for the question.

The classifier for the softmax loss is only used during training. At inference, when computing sentence similarities, it is not used. You can use there for example cosine similarity or Euclidean distance.

See this example:
https://github.com/UKPLab/sentence-transformers/blob/master/examples/training_nli_bert.py

There, BERT is fine-tuned with the softmax loss. However, for the dev and test set, cosine-similarity based on the sentence embeddings outputs is used.

Hence, finding the pair with the highest similarity only requires n=10 000 sentence embeddings computations and then running cosine similarity to compare all possible pairs.

Best
-Nils Reimers

**Comment by EdwardMao** [comment]
> Hi @EdwardMao
> thanks for the question.
> 
> The classifier for the softmax loss is only used during training. At inference, when computing sentence similarities, it is not used. You can use there for example cosine similarity or Euclidean distance.
> 
> See this example:
> https://github.com/UKPLab/sentence-transformers/blob/master/examples/training_nli_bert.py
> 
> There, BERT is fine-tuned with the softmax loss. However, for the dev and test set, cosine-similarity based on the sentence embeddings outputs is used.
> 
> Hence, finding the pair with the highest similarity only requires n=10 000 sentence embeddings computations and then running cosine similarity to compare all possible pairs.
> 
> Best
> -Nils Reimers

Hi,
It seems strange to me that the embedding trained with softmax could be evaluated by cosine similarity. Would it be better if you train the model with cosine similarity loss? Otherwise, the information in the classify layer is lost.
BTW, another question is that have you evaluate the performance of bert-as-service, which is also a model that derives sentence embedding throught BERT? What is the difference between these two models? 
This is a wonderful work and thank you for your answer!
Best,

**Comment by nreimers** [comment]
Hi @EdwardMao ,
I agree that it is a bit strange. I think the first paper working with this setup was the InferSent paper. They also trained on NLI data with a softmax classifier.

As the classifier is rather simple on works on the representation of (u, v |u-v|), the vector space must encode semantic properties, for example, that similar sentences are close in that space.

If you have similarity data,  than of course cosine similarity is better. See Table 2 in the Sentence-BERT paper (https://arxiv.org/pdf/1908.10084.pdf).

On NLI data (with softmax loss), BERT-base achieves a performance of 77 (on STS benchmark). If trained on the training set of STS benchmark, it achieves a performance of 84.

I also evaluated bert-as-a-service. See Table 1 in the Sentence-BERT paper.

Averaging BERT output vectors or using the CLS token, as it is done by bert-as-a-service, generates rather bad sentence embeddings. These sentence embeddings are often worse than averaging GloVe embeddings.

Best,
Nils Reimers

**Comment by EdwardMao** [comment]
> Hi @EdwardMao ,
> I agree that it is a bit strange. I think the first paper working with this setup was the InferSent paper. They also trained on NLI data with a softmax classifier.
> 
> As the classifier is rather simple on works on the representation of (u, v |u-v|), the vector space must encode semantic properties, for example, that similar sentences are close in that space.
> 
> If you have similarity data, than of course cosine similarity is better. See Table 2 in the Sentence-BERT paper (https://arxiv.org/pdf/1908.10084.pdf).
> 
> On NLI data (with softmax loss), BERT-base achieves a performance of 77 (on STS benchmark). If trained on the training set of STS benchmark, it achieves a performance of 84.
> 
> I also evaluated bert-as-a-service. See Table 1 in the Sentence-BERT paper.
> 
> Averaging BERT output vectors or using the CLS token, as it is done by bert-as-a-service, generates rather bad sentence embeddings. These sentence embeddings are often worse than averaging GloVe embeddings.
> 
> Best,
> Nils Reimers

It occurs to me that the classify layer makes sense. Let's say we use NLI as our fine-tuning dataset. If we only use cosine similarity loss as training loss, the sentence embedding would be only for NLI task. However, when use classify layer, many information about NLI will be stored in the classify layer and many semantic information will be stored in BERT embedding, which would be more universal for other tasks. That's my understanding.
When I dive into the code of binaryembeddingsimilarityevaluator.py, the cosine_middle is set to be the mean of all cosine scores. Could you please tell me why use cosine_middle? Why not use 0 or  any other constant? Because it would perform bad if the dev dataset is all True or all False.
Thank you!

**Comment by nreimers** [comment]
You cannot use cosine similarity for the NLI dataset. NLI has 3 labels: entailment, contradition, neutral. So the only way to model this is with a softmax classifier.

binaryembeddingsimilarityevaluator.py is just a helping script to evaluate the quality for sentence embeddings. It assumes that you have a balanced binary dataset, i.e. 50% of the pairs are similar (label=1) and 50% of the pairs are dissimilar (label=0). It then check using cosine-similarity if the sentence embedding model can distinguish the 50% similar pairs from the 50% dissimilar pairs. The evaluator is not intended for sets where every pair is label=1 or label=0.

Why not use 0 or a constant: Because you cannot tell a-priori where you should put the threshold. You can use either a dev set to find a decision boundary for similar / dissimilar pairs, or, as in this case, assume that 50% or the pairs are similar / dissimilar and then use the median.

For production this evalutor is of course quite useless. But it helps during training to assess some sentence embeddings models.
